docs/design/2020-05-08-cluster-index.md

```diff
@@ -0,0 +1,254 @@
+
+#### On Point Select
+
+If the primary key is a single integer column,
+We construct the row key by the primary key column value directly, do a lookup on TiKV.
+
+Otherwise, we need to construct the index key with the primary key value, read the handle for the index entry,
+Then construct the row key and do another lookup on TiKV.
+
+#### On Range Scan
+If the primary key is a single integer column,
+We construct the range of the row keys directly, do a single scan on TiKV.
+
+Otherwise, we need to construct the index range, scan index entries on TiKV to collect the handle,
+Then do many point lookups On TiKV to read the rows.
+
+## Proposal
+
+Support clustered index for not only single integer column primary key but all kinds of primary key. 
+
+### the Handle interface
+```go
+package kv
+// Handle is the ID of a row.
+type Handle interface {
+   // IsInt returns if the handle type is int64.
+   IsInt() bool
+   // IntValue returns the int64 value if IsInt is true, it panics if IsInt returns false.
+   IntValue() int64
+   // Next returns the minimum handle that is greater than this handle.
+   Next() Handle
+   // Equal returns if the handle equals to another handle, it panics if the types are different.
+   Equal(h Handle) bool
+   // Compare returns the comparison result of the two handles, it panics if the types are different.
+   Compare(h Handle) int
+   // Encoded returns the encoded bytes.
+   Encoded() []byte
+   // Len returns the length of the encoded bytes.
+   Len() int
+   // NumCols returns the number of columns of the handle,
+   NumCols() int
+   // EncodedCol returns the encoded column value at the given column index.
+   EncodedCol(idx int) []byte
+   // String implements the fmt.Stringer interface.
+   String() string
+}
+
+// IntHandle implements the Handle interface for the int64 type handle.
+type IntHandle int64
+
+// CommonHandle implements the Handle interface for non-int64 type handle.
+type CommonHandle struct {
+   encoded       []byte
+   colEndOffsets []uint16
+}
+
+// HandleMap is the map for Handle.
+type HandleMap struct {
+   ints map[int64]interface{}
+   strs map[string]strHandleVal
+}
+
+type strHandleVal struct {
+   h   Handle
+   val interface{}
+}
+```
+To extend the handle type, we need to abstract the handle to an interface that has the
+needed methods. 
+
+There are two concrete types: IntHandle and CommonHandle.
+
+If the primary key is a single column integer, the handle type would be IntHandle.
+Otherwise, the handle type would be CommonHandle.
+
+handles are used as map key in many places, one way to adapt to it is to use the encoded bytes as the key, but that would introduce memory allocation for IntHandle, so we use a HandleMap type that has two maps for the two handle types to avoid the memory allocation for IntHandle.
+
+Integer handle row key and index key formats are the same as the current implementation.
+We only need to care about the common handle type.
+
+### Codec
```

Comment:
I think we should also care about the `UnCommitIndexKVFlag` on values in memdb for `union scan` optimization. e.g.
https://github.com/pingcap/tidb/blob/6814f8c94b1d239f6be66a33cf3a7aa11d9f1f12/table/tables/index.go#L335

Reply:
That would be too detailed for the proposal, it is handled in the PR https://github.com/pingcap/tidb/pull/16998